### PR TITLE
[SYCL] Raise bit_cast to SYCL namespace

### PR DESCRIPTION
**Pull request by cperkinsintel**
SYCL2020 has bit_cast provided in SYCL namespace now. Moved it and updated tests and internal references

Signed-off-by: Chris Perkins <chris.perkins@intel.com>

### CODE REVIEW

File: sycl/ReleaseNotes.md
Comment (bader) [inline sycl/ReleaseNotes.md:766]:
```suggestion
    as a `sycl::detail::bit_cast` [e3da4ef]
```

File: sycl/include/CL/sycl/detail/helpers.hpp
Comment (bader) [inline sycl/include/CL/sycl/detail/helpers.hpp:248]: Probably the definition to be moved from `sycl/include/CL/sycl/detail/helpers.hpp` to dedicated `sycl/include/CL/sycl/bit_cast.hpp` header.

File: sycl/ReleaseNotes.md
Comment (AlexeySachkov) [reply sycl/ReleaseNotes.md:766]: I'm not sure that we should change this at all. These are for past releases: there is even commit hashes specified at line 758

Comment (Pennycook) [reply sycl/ReleaseNotes.md:766]: This made me wonder why we still have the  draft `SYCL_INTEL_bitcast` extension documentation at all...  We never implemented it (because what we did implement was in `sycl::detail::`), and what we're implementing now is not that extension but SYCL 2020.  Can we remove it?

Would be interested to hear from @pvchupin and other @intel/dpcpp-specification-reviewers on this.

Comment (cperkinsintel) [reply sycl/ReleaseNotes.md]: restored. thanks

File: sycl/include/CL/sycl/detail/helpers.hpp
Comment (cperkinsintel) [reply sycl/include/CL/sycl/detail/helpers.hpp]: thanks.  Moved it.

**Comment by cperkinsintel** [comment]
I believe the Jenkins Precommit stage is failing because there is a copy of the `bit_cast` test over there.  I have a draft PR ( https://github.com/intel/llvm-test-suite/pull/228 ) that updates the `bit_cast` test on llvm-test-suite. It should be able to pass once this PR is merged.

File: sycl/include/CL/sycl/bit_cast.hpp
Comment (bader) [inline sycl/include/CL/sycl/bit_cast.hpp:1]: helpers -> bit_cast

File: sycl/include/CL/sycl/bit_cast.hpp
Comment (bader) [inline sycl/include/CL/sycl/bit_cast.hpp:56]: Please, add a new line at the end of the file.

**Review by bader — DISMISSED** [review]

File: sycl/include/CL/sycl/stl.hpp
Comment (romanovvlad) [inline sycl/include/CL/sycl/stl.hpp:15]:
```suggestion
#include <CL/sycl/bit_cast.hpp>

```

Comment (cperkinsintel) [reply sycl/include/CL/sycl/stl.hpp]: done

File: sycl/include/CL/sycl/stl.hpp
Comment (romanovvlad) [inline sycl/include/CL/sycl/stl.hpp:15]:
```suggestion
#include <CL/sycl/bit_cast.hpp>

```

**Comment by romanovvlad** [comment]
> failing because there is a copy of the ``

Suggest having `sycl::detail::bit_cast` available by having `using bit_cast = sycl::bit_cast` in the `detail` namespace.

File: sycl/include/CL/sycl/bit_cast.hpp
Comment (bader) [inline sycl/include/CL/sycl/bit_cast.hpp:46]: Just wondering: shouldn't this condition be checked regardless compiler built-ins support?

**Review by bader — APPROVED** [review]

**Review by romanovvlad — APPROVED** [review]

File: sycl/include/CL/sycl/bit_cast.hpp
Comment (cperkinsintel) [reply sycl/include/CL/sycl/bit_cast.hpp:46]: I did not write this code, I just moved it out of helpers.hpp.  However, I think the answer is that that std::bit_cast only requires `is_trivially_copyable`.  But, if you are implementing it yourself with `memcpy` that implementation will also require `is_trivially_default_constructible`.  The code in the "Notes" section here suggests this: https://en.cppreference.com/w/cpp/numeric/bit_cast .

Comment (bader) [reply sycl/include/CL/sycl/bit_cast.hpp:46]: Actually it says that additional requirement is `is_trivially_constructible`, not `is_trivially_default_constructible`.
And our implementation requires `is_trivially_constructible` even if `std::bit_cast` is supported. According to the cppreference it's not necessary...

Comment (bader) [reply sycl/include/CL/sycl/bit_cast.hpp:46]: Sorry, I mixed up `is_trivially_copyable` and `is_trivially_constructible` in my previous comment. Please, ignore it.

**Comment by cperkinsintel** [comment]
ping to code owners. This should be ready.

**Comment by pvchupin** [comment]
That's @AlexeySachkov

**Review by AlexeySachkov — APPROVED** [review]

**Pull request merged**
